source/common/grpc/grpc_web_filter.cc

```diff
@@ -0,0 +1,127 @@
+#include "common/grpc/grpc_web_filter.h"
+
+#include <arpa/inet.h>
+
+#include "common/common/base64.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Grpc {
+
+// Bit mask denotes a trailers frame of gRPC-Web.
+const uint8_t GrpcWebFilter::GRPC_WEB_TRAILER = 0b10000000;
+
+// Implements StreamDecoderFilter.
+// TODO(fengli): Implements the subtypes of gRPC-Web content-type, like +proto, etc.
+Http::FilterHeadersStatus GrpcWebFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == content_type->value().c_str()) {
+    // Checks whether gRPC-Web client is sending base64 encoded request.
+    is_text_request_ = true;
+  }
+  headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Grpc);
+
+  const Http::HeaderEntry* accept = headers.get(Http::Headers::get().Accept);
+  if (accept != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == accept->value().c_str()) {
+    // Checks whether gRPC-Web client is asking for base64 encoded response.
+    is_text_response_ = true;
+  }
+
+  // Adds te:trailers to upstream HTTP2 request. It's required for gRPC.
+  headers.addStatic(Http::Headers::get().TE, Http::Headers::get().TEValues.Trailers);
```

Comment:
Will address in a follow up PR. Also, there are other stuffs need to be enhanced, like add some wrapper methods in Buffer::OwnedImpl to avoid the additional buffer allocation. Today I have to use a temp buffer when processing trailers, as the Buffer::OwnedImpl doesn't allow me to insert data to a particular position.